keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,430 @@
+is to let the initiating actor such as VPA handle the failure. Alternately, a
+Controller can either retry the update, or reschedule the Pod based on policy.
+
+PodSpec is extended with a new flag, PodSpec.RetryPolicy, with possible values:
+* NoRetry - the default value; do nothing, initiating actor handles failure,
+* RetryUpdate - Controller retries resource update in-place,
+* Reschedule - Controller evicts Pod, and creates updated Pod for scheduling.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter is changed to *runtimeapi.ContainerResources*,
+that is runtime agnostic.
+
+### Flow Control
+
+Presently, when a Pod is created, Scheduler is responsible for selecting a suitable
+Node that accommodates the Pod. In order to allow Pod resizing operation to work
+well with new Pod scheduling workflow, this design keeps Scheduler in the Pod
+resizing decision path.
+
+The following steps denote a typical flow of an in-place resize operation for a Pod
+with ResizePolicy set to Update for all its Containers.
+
+1. The initiating actor updates ResourceRequirements using PATCH verb.
+1. API Server validates the new ResourceRequirements
+   (e.g. limits are not below requested resources, QoS class does not change).
+1. API Server calls all Admission Controllers to verify the Pod Update.
+   * If any of the Controllers reject the update, they set PodResizeSuccess
+     PodCondition's Reason to Rejected<Reason>, and set Status to false.
+1. API Server updates PodSpec object with the new desired ResourceRequirements.
+1. Scheduler observes that ResourceRequirements and ResourceBindings differ.
+   It checks its cache to determine if in-place resource resizing is possible.
+   * If Node has capacity to accommodate new resource values, it updates
+     ResourceBindings with desired values from ResourceRequirements, updates
+     its resource cache using max(ResourceRequirements, ResourcesAllocated),
+     and sets PodResourcesBound PodCondition's Status to true, and Reason to
+     Accepted.
+   * If required, it pre-empts lower-priority Pods, setting the PodResourcesBound
+     PodCondition's Status to false, and Reason to AwaitingPreemption. Once
+     lower-priority Pods are evicted, Scheduler sets PodResourcesBound
+     PodCondition's Status to true, Reason to Accepted, and the flow continues.
+   * If Node does not have capacity to accommodate the new resource values, it
+     sets the PodResourcesBound PodCondition's Reason to FailedNodeCapacity, and
+     sets Status to false.
+1. Kubelet observes that ResourceBindings have changed, and checks its Node
+   allocatable resources against the new ResourceBindings for fit.
+   * Kubelet sees that new ResourceBindings fits, applies the new cgroup limits
+     to the Pod and its Containers, and sets PodStatus ResourcesAllocated to
```

Comment:
on scale down, it will only set this when it observes that the memory limit is reduced and accepted by the bounding cgroup?  will the kubelet need to make an eviction decision of its own if its not able to force reclaim?

Reply:
Do you mean evict the Pod that is being resized? If yes, I don't think so - restarting the Pod in-place should suffice.

Reply:
Kubelet eviction is not in this KEP, a possible future extension.
We now update Spec ResourcesAllocated on accepting the resize, and update Status Resources upon updating limits. Hopefully this addresses the questions above. Resolving this issue, please reopen if unresolved.